libc/src/__support/common.h

```diff
@@ -25,7 +25,7 @@
 #define LLVM_LIBC_FUNCTION_IMPL(type, name, arglist)                           \
   LLVM_LIBC_FUNCTION_ATTR decltype(LIBC_NAMESPACE::name)                       \
       __##name##_impl__ __asm__(#name);                                        \
-  decltype(LIBC_NAMESPACE::name) name [[gnu::alias(#name)]];                   \
```

Comment:
The semantic of these 2 are different.  In the old one,
```
decltype(LIBC_NAMESPACE::name) name [[gnu::alias(#name)]];
```
will give you a public C++ function `LIBC_NAMESPACE::name` (b/c we used this macro inside `namespace LIBC_NAMESPACE { ... }` which is internally aliased to the public C symbol `#name`.
On the other hand, the second one using `asm` will simply make the public symbol of `LIBC_NAMESPACE::name` as unmangled `#name`.

So this change will make `LIBC_NAMESPACE::name` not available outside of the translation unit.

Reply:
How does this macro result in two symbols in the emitted binary, one non-mangled, and one mangled wrt. the namespace?  If I extract out this macro, I only see the non-mangled name.

https://godbolt.org/z/oedsKxroP

But for full build, if you run `llvm-readelf -s <build dir>/libc/lib/libc.a` you see the pairs of symbol names.

There's something else I'm missing here...

Reply:
So currently, the unmangled symbol `#name` is declared as the public symbol of `LIBC_NAMESPACE::name_impl`, which is declared in https://github.com/llvm/llvm-project/blob/main/libc/src/__support/common.h#L27 and define in https://github.com/llvm/llvm-project/blob/main/libc/src/__support/common.h#L29.

https://github.com/llvm/llvm-project/blob/main/libc/src/__support/common.h#L28 then defines the other C++ public symbol `LIBC_NAMESPACE::name` and aliases it with the unmangled C symbol.